### PR TITLE
THRIFT-5237 Implement MAX_MESSAGE_SIZE and consolidate limits into a TConfiguration class (c_glib)

### PR DESCRIPTION
**Pull request by zeshuai007**
TConfiguration class (c_glib)

  

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [ ] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

**Comment by Jens-G** [comment]
I'd recommend to ask on the mailing list to find someone to review. I can have a look but my c_glib knowledge is far from being an expert, so a second opinion would be sort of required anyway.

File: lib/c_glib/src/thrift/c_glib/protocol/thrift_binary_protocol.c [hunk 1]
```diff
@@ -557,6 +561,14 @@ thrift_binary_protocol_read_map_begin (ThriftProtocol *protocol,
     return -1;
   }
 
+  if(FALSE == ttc->checkReadBytesAvailable(THRIFT_TRANSPORT(tp->transport),
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/protocol/thrift_binary_protocol.c:564]: `if( FALSE == ...)`  - we can do better, can we?

File: lib/c_glib/src/thrift/c_glib/protocol/thrift_binary_protocol.c [hunk 2]
```diff
@@ -603,6 +618,13 @@ thrift_binary_protocol_read_list_begin (ThriftProtocol *protocol,
...
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/protocol/thrift_binary_protocol.c:621]: Same here

File: lib/c_glib/src/thrift/c_glib/protocol/thrift_compact_protocol.c [hunk 3]
```diff
@@ -1106,6 +1108,15 @@ thrift_compact_protocol_read_map_begin (ThriftProtocol *protocol,
     return -1;
   }
 
+  
+  if(FALSE == ttc->checkReadBytesAvailable(THRIFT_TRANSPORT (tp->transport),
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/protocol/thrift_compact_protocol.c:1112]: again, and there are a few more below

File: lib/c_glib/src/thrift/c_glib/transport/thrift_buffered_transport.c [hunk 4]
```diff
@@ -379,6 +424,36 @@ thrift_buffered_transport_class_init (ThriftBufferedTransportClass *cls)
...
+  g_object_class_install_property (gobject_class,
+		                   PROP_THRIFT_BUFFERED_TRANSPORT_CONFIGURATION,
+				   param_spec);
+
+  param_spec = g_param_spec_long ("remainingmessagesize",
+		                  "remainingmessagesize (construct)",
+				  "Set the size of the remaining message",
+				  0, /* min */
+                                  G_MAXINT32, /* max */
+				  10485760, /* default by construct */
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_buffered_transport.c:441]: can we put that `10485760` into some constant and give it a meaningful name?

File: lib/c_glib/src/thrift/c_glib/transport/thrift_buffered_transport.c [hunk 5]
```diff
@@ -379,6 +424,36 @@ thrift_buffered_transport_class_init (ThriftBufferedTransportClass *cls)
...
+  g_object_class_install_property (gobject_class,
+		                   PROP_THRIFT_BUFFERED_TRANSPORT_REMAINING_MESSAGE_SIZE,
+				   param_spec);
+
+  param_spec = g_param_spec_long ("knowmessagesize",
+		                  "knowmessagesize (construct)",
+				  "Set the size of the know message",
+				  0, /* min */
+                                  G_MAXINT32, /* max */
+				  10485760, /* default by construct */
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_buffered_transport.c:452]: same, plus some more below

File: lib/c_glib/src/thrift/c_glib/transport/thrift_server_socket.c [hunk 6]
```diff
@@ -246,6 +270,20 @@ thrift_server_socket_set_property (GObject *object, guint property_id,
     case PROP_THRIFT_SERVER_SOCKET_BACKLOG:
       socket->backlog = g_value_get_uint (value);
       break;
+    case PROP_THRIFT_SERVER_SOCKET_CONFIGURATION:
+      transport->configuration = g_value_dup_object (value);
+      if(transport->configuration != NULL)
+      {
+	transport->remainingMessageSize_ = transport->configuration->maxMessageSize_;
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_server_socket.c:277]: minor: indent?

File: lib/c_glib/src/thrift/c_glib/transport/thrift_server_transport.c [hunk 7]
```diff
@@ -21,15 +21,216 @@
...
+
+    THRIFT_UNUSED_VAR (pspec);
+
+    switch (property_id)
+    {
+        case PROP_THRIFT_SERVER_TRANSPORT_CONFIGURATION:
+          transport->configuration = g_value_get_object (value);
+	  if(transport->configuration != NULL)
+	  {
+              transport->remainingMessageSize_ = transport->configuration->maxMessageSize_;
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_server_transport.c:172]: minor: indent?

File: lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c [hunk 8]
```diff
@@ -129,17 +141,209 @@ thrift_transport_real_read_all (ThriftTransport *transport, gpointer buf,
...
+
+gboolean
+thrift_transport_resetConsumedMessageSize(ThriftTransport *transport, glong newSize, GError **error)
+{
+    ThriftTransport *tt = THRIFT_TRANSPORT (transport);
+    if(newSize < 0)
+    {
+	if(tt->configuration != NULL)
+	{
+            tt->knowMessageSize_ = tt->configuration->maxMessageSize_;
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c:188]: minor: indent?

File: lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c [hunk 9]
```diff
@@ -129,17 +141,209 @@ thrift_transport_real_read_all (ThriftTransport *transport, gpointer buf,
...
+{
+    ThriftTransport *tt = THRIFT_TRANSPORT (transport);
+    if(tt->remainingMessageSize_ > numBytes)
+    {
+        tt->remainingMessageSize_ -= numBytes;
+    }
+    else
+    {
+        tt->remainingMessageSize_ = 0;
+	g_set_error(error,
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c:225]: indent?

File: lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c [hunk 10]
```diff
@@ -129,17 +141,209 @@ thrift_transport_real_read_all (ThriftTransport *transport, gpointer buf,
...
+  GObjectClass *gobject_class = G_OBJECT_CLASS (cls);
+  GParamSpec *param_spec = NULL;
+  
+  /* setup accessors and mutators */
+  gobject_class->get_property = thrift_transport_get_property;
+  gobject_class->set_property = thrift_transport_set_property;
+  
+  param_spec = g_param_spec_object ("configuration",
+		                    "configuration (construct)",
+				    "Thtift Configuration",
```
Comment (Jens-G) [inline lib/c_glib/src/thrift/c_glib/transport/thrift_transport.c:316]: "Thtift Configuration" => "Thrift Configuration"

**Comment by zeshuai007** [comment]
Done.Please review again.

**Comment by cyril867** [comment]
@zeshuai007  hi, Does this merge fix CVE-2020-13949 ?

**Comment by dzhang-b** [comment]
Hi, is this fix merged into 0.14? Could the fix backported to older version like 0.11, 0.12? Thanks

**Pull request closed without merging**